### PR TITLE
Add relative paths check

### PR DESCRIPTION
**Pull request by ostolero**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [link to the issue](https://jira-hq.paloaltonetworks.local/browse/CIAC-654)

## Description
Add validation to identify relative urls in README files.

## Must have
- [x] Tests
- [ ] Documentation
- [ ] [VS Code Extension](https://docs.gitlab.com/ee/ci/yaml/#tags)
  - [ ] Functionality implemented in the extension - <u>\<link-to-pr-here\></u>
  - [ ] N/A - Functionality cannot be implemented in the extension because <u>\<add-reason-here\></u>

### CODE REVIEW

**Comment by ostolero** [comment]
Working on fixing the content errors. :)

File: demisto_sdk/commands/common/hook_validations/readme.py [hunk 1]
```diff
@@ -240,7 +241,41 @@ def verify_readme_image_paths(self) -> bool:
...
+            is_pack_readme (bool) - True if the the README file is a pack README, default: False
+
+        Returns:
+            list: List of the errors found
+        """
+        error_list = []
+        # If error was found, print it only if its not a pack readme. For pack readme, the PackUniqueFilesValidator
+        # class handles the errors and printing.
+        should_print_error = not is_pack_readme
+        relative_urls = re.findall(r'^(?![!])(\[.*?\])\(((?![http|#|mailto:]).*?)\)$', self.readme_content,
```
[the remark this answers is not in the document]
Comment (JudahSchwartz) [reply demisto_sdk/commands/common/hook_validations/readme.py:259]: whats the case of adding ! to override this? `^(?![!])`
lets add a test for that

File: demisto_sdk/tests/test_files/README-invalid.md [hunk 2]
```diff
@@ -1,7 +1,10 @@
 Use the Quin-C AccessData integration to protect against and provide additional visibility into phishing and other malicious email attacks. This integration was integrated and tested with version 20190926 of Quin-C Accessdata.
 
 Documentation for the integration was provided by Quin-C.
-
+For more info check out think links-
```
Comment (JudahSchwartz) [inline demisto_sdk/tests/test_files/README-invalid.md:4]: please also add tests for html. I think its no issue to be in the same file

**Review by JudahSchwartz — APPROVED** [review]
Very nice!

File: demisto_sdk/commands/common/hook_validations/readme.py (hunk 1 above)
Comment (JudahSchwartz) [inline demisto_sdk/commands/common/hook_validations/readme.py:259]: this regex seems not to work. Check this case
[Good link](hello.co.il)
`^(?![!])(\[.*?\])\(((?!(?:https?:\/\/)|#|(?:mailto:)).*?)\)$`

File: demisto_sdk/tests/test_files/Packs/DummyPack2/README.md [hunk 3]
```diff
@@ -2,6 +2,11 @@ PhishTank is a free community site where anyone can submit, verify, track and sh
 This integration was integrated and tested with version 1.0.1 of PhishTank.
 ## Configure PhishTankV2 on Cortex XSOAR
 
+### Relative urls test-
+[Good link](htttps://www.good.co.il)
```
[the remark this answers is not in the document]
Comment (ostolero) [reply demisto_sdk/tests/test_files/Packs/DummyPack2/README.md:6]: Oops, that is a typo I changes to the Regex you suggested and that fixed the issue :). Added a test for https://example.com/. btw this link is valid.

**Comment by ostolero** [comment]
Both files failing in the validate-files test are Partner contributions.
Waiting for the partners to respond.

File: demisto_sdk/commands/common/hook_validations/readme.py [hunk 4]
```diff
@@ -28,8 +28,10 @@
     get_url_with_retries, get_yaml, get_yml_paths_in_dir, print_warning,
     run_command_os)
 
-json = JSON_Handler()
+RELATIVE_HREF_URL_REGEX = r'(<.*?href\s*=\s*"((?!(?:https?:\/\/)|#|(?:mailto:)).*?)")'
+RELATIVE_URL_REGEX = r'^(?![!])(\[.*?\])\(((?!(?:https?:\/\/)|#|(?:mailto:)).*?)\)$'
```
Comment (JudahSchwartz) [inline demisto_sdk/commands/common/hook_validations/readme.py:32]:
```suggestion
RELATIVE_MARKDOWN_URL_REGEX = r'^(?![!])(\[.*?\])\(((?!(?:https?:\/\/)|#|(?:mailto:)).*?)\)$'
```

File: demisto_sdk/tests/test_files/README-invalid.md (hunk 2 above)
Comment (JudahSchwartz) [reply demisto_sdk/tests/test_files/README-invalid.md:4]: I would add a couple more html tests, like you have for markdown

File: demisto_sdk/commands/common/tests/readme_test.py [hunk 5]
```diff
@@ -74,6 +74,36 @@ def test_are_modules_installed_for_verify_false_res(tmp_path):
...
+    captured_output = io.StringIO()
+    sys.stdout = captured_output  # redirect stdout.
+    absolute_urls = ["https://www.good.co.il", "https://example.com", "https://github.com/demisto/content/blob/123"
+                                                                      "/Packs/FeedOffice365/doc_files/test.png"]
+    relative_urls = ["relative1.com", "www.relative2.com", "hreftesting.com"]
+    readme_validator = ReadMeValidator(INVALID_MD)
+    result = readme_validator.verify_readme_relative_urls()
+    sys.stdout = sys.__stdout__  # reset stdout.
+    assert not result
+    assert absolute_urls[0] not in captured_output.getvalue()
```
Comment (JudahSchwartz) [inline demisto_sdk/commands/common/tests/readme_test.py:99]: loop?

File: demisto_sdk/commands/common/tests/pack_unique_files_test.py [hunk 6]
```diff
@@ -644,6 +644,35 @@ def test_validate_pack_readme_invalid_images(self):
...
+        errors = self.validator.get_errors()
+        assert not result
+        assert 'Relative urls are not supported within README, If this is not a relative url make sure you add ' \
+               'https:// at the start of it:\nrelative1.com. ' in errors
+        assert 'Relative urls are not supported within README, If this is not a relative url make sure you add ' \
+               'https:// at the start of it:\nwww.relative2.com. ' in errors
+        assert 'Relative urls are not supported within README, If this is not a relative url make sure you add ' \
+               'https:// at the start of it:\nhreftesting.com' in errors
+        # this path is not an image path and should not be shown.
+        assert 'htttps://www.good.co.il' not in errors
```
Comment (JudahSchwartz) [inline demisto_sdk/commands/common/tests/pack_unique_files_test.py:672]:
```suggestion
        assert 'https://www.good.co.il' not in errors
```

**Comment by ShahafBenYakir** [comment]
@ostolero @JudahSchwartz In both of the cases it looks like a false alarm actually
The Chronicle one clearly shows that we are flagging empty URLs where we shouldn't.
Reg cloudshark it is not really a relative URL so I think we can ignore it.

File: demisto_sdk/commands/common/errors.py [hunk 7]
```diff
@@ -1785,6 +1786,12 @@ def branch_name_in_readme_image_absolute_path_error(path):
     def invalid_readme_insert_image_link_error(path):
         return f'Image link was not found, either insert it or remove it:\n{path}'
 
+    @staticmethod
+    @error_code_decorator
+    def invalid_readme_relative_url_error(path):
+        return f'Relative urls are not supported within README, If this is not a relative url make sure you add ' \
+               f'https:// at the start of it:\n{path}. '
```
Comment (dorschw) [inline demisto_sdk/commands/common/errors.py:1793]:
```suggestion
        return f'Relative urls are not supported within README. If this is not a relative url, please add  ' \
               f'an https:// prefix:\n{path}. '
```

File: demisto_sdk/commands/common/hook_validations/readme.py [hunk 8]
```diff
@@ -240,6 +243,47 @@ def verify_readme_image_paths(self) -> bool:
...
+            list: List of the errors found
+        """
+        error_list = []
+        # If error was found, print it only if its not a pack readme. For pack readme, the PackUniqueFilesValidator
+        # class handles the errors and printing.
+        should_print_error = not is_pack_readme
+        relative_urls = re.findall(RELATIVE_URL_REGEX, self.readme_content,
+                                   re.IGNORECASE | re.MULTILINE)
+        relative_urls += re.findall(RELATIVE_HREF_URL_REGEX, self.readme_content,
+                                    re.IGNORECASE | re.MULTILINE)
```
Comment (dorschw) [inline demisto_sdk/commands/common/hook_validations/readme.py:276]: consider using a precompiled regex object (`re.compile(..)`) (nice to have, not a must)

File: demisto_sdk/tests/test_files/README-invalid.md (hunk 2 above)
Comment (ostolero) [reply demisto_sdk/tests/test_files/README-invalid.md:4]: Added a couple more, the html files are old and I don't think we create more so I think it should be enough.

**Comment by ostolero** [comment]
@ShahafBenYakir @JudahSchwartz Shouldn't empty links be caught in our test? they're not exactly relative but are invalid. The relative image test catches empty links as well.
The CloudaShark is some output of docs generated I think, It should be fixed if possible.
Will the failed test affect our nightly or other routine tests if we force merge?

**Comment by ShahafBenYakir** [comment]
>The CloudaShark is some output of docs generated I think, It should be fixed if possible.

The integration command outputs MD link so this is what appears in the HTML, you can obviously remove it from the HTML.
Also, I don't think the same validation should catch empty links, we can consider adding another validation if needed.

File: demisto_sdk/commands/common/tests/pack_unique_files_test.py [hunk 9]
```diff
@@ -644,6 +644,41 @@ def test_validate_pack_readme_invalid_images(self):
...
+                    - Invalid absolute paths were not caught.
+                    - Image paths were not caught
+        """
+        self.validator = PackUniqueFilesValidator(os.path.join(self.FILES_PATH, 'DummyPack2'))
+
+        result = self.validator.validate_pack_readme_relative_urls()
+        errors = self.validator.get_errors()
+        relative_error = 'Relative urls are not supported within README. If this is not a relative url, please add ' \
+                         'an https:// prefix:\n'
+        assert not result
```
Comment (JudahSchwartz) [inline demisto_sdk/commands/common/tests/pack_unique_files_test.py:666]: I think using loops here would be clearer to. For all urls, assert ...

**Review by JudahSchwartz — APPROVED** [review]
Besides for refactoring the method I mentioned in the other PR, this looks good to go! 
(Notice the other comment I left here)
Great job!

**Comment by JudahSchwartz** [comment]
@ostolero @edik24  are we able to add the error code here to the pack ignore of that contributor?

**Comment by edik24** [comment]
@JudahSchwartz sorry, missed your comment.
I'm not sure what you mean by adding the error code.

anyway,
the status of chronicle is that they anyway intend to open PR this week to fix something in their pack, so they will fix the links as well - discussed it with them already.

regarding cloudhsark, I sent them email, didn't hear back from them so far.
in their case the readme is html and the one broken link is markdown, this is why it's broken - so if we won't hear from them, maybe we can fix it by ourselves?
@ostolero it's possible?

**Comment by ostolero** [comment]
Closing this, these changes will be merged as part of https://github.com/demisto/demisto-sdk/pull/2179

**Pull request closed without merging**